### PR TITLE
Configure auto-send and auto-delete at the form level

### LINKED ISSUE #791: Add form level support for auto-send and auto-delete
**Problem**
We have a user whose campaign has forms that need to be deleted on send because they are sensitive. The same user has other forms that need to be kept on the device because the data is useful to the day to day work of the enumerator.

**Proposal**
In the settings sheet of the XLSForm, we will add an `auto_send` and `auto_delete` heading. If the values underneath this heading is `true`, then the form will behave accordingly and override the settings of the device. In Collect, the XML we'll look for is: `<submission orx:auto-delete="true" orx:auto-submit="true" >`

### PR DESCRIPTION
**Pull request by grzesiek2010**
This PR is in reference to resolve #791

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java [hunk 1]
```diff
@@ -176,6 +176,22 @@ private void uploadForms(Context context) {
         }
     }
 
+    private boolean submitInstance(String jrFormId, boolean autoSendSettings) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java]: I would call this something like `isFormAutoSendEnabled` to clarify what it does. I think that will make the code that uses it easier to read.

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java]: Also, it would be useful to have some comments that indicate that if the form explicitly sets the auto-send property, then it overrides the preferences. The general settings value is only used if auto-send isn't set at the form level. That will make it easier to verify the intent of the code in the future.

File: collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java [hunk 2]
```diff
@@ -115,8 +113,10 @@ private void uploadForms(Context context) {
                 if (c != null && c.getCount() > 0) {
                     c.move(-1);
                     while (c.moveToNext()) {
-                        Long l = c.getLong(c.getColumnIndex(InstanceColumns._ID));
-                        toUpload.add(l);
+                        if (submitInstance(c.getString(c.getColumnIndex(InstanceColumns.JR_FORM_ID)), interfaceIsEnabled(context, networkInfo))) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java]: This is hard to parse. I think using a variable for the formID would help.

File: collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java [hunk 3]
```diff
@@ -80,9 +80,7 @@ public void onReceive(Context context, Intent intent) {
             if (ni == null || !ni.isConnected()) {
                 // not connected, do nothing
             } else {
-                if (interfaceIsEnabled(context, ni)) {
-                    uploadForms(context);
-                }
+                uploadForms(context, ni);
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/receivers/NetworkReceiver.java]: Instead of passing the network interface in here, I think it would be better to pass in the result of `interfaceIsEnabled`. That value doesn't change across all the forms so it's better if it doesn't have to be recomputed.

File: collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java [hunk 4]
```diff
@@ -689,4 +683,24 @@ private static String getFileExtension(String fileName) {
         }
         return fileName.substring(dotIndex + 1).toLowerCase(Locale.ROOT);
     }
+
+    private boolean deleteInstance(String jrFormId) {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java]: How about `shouldDeleteInstance` or `isAutoDeleteEnabled` so it's clear it's a boolean method? For symmetry with `isFormAutoSendEnabled` above, you could also pass this value in.

File: collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java [hunk 5]
```diff
@@ -689,4 +683,24 @@ private static String getFileExtension(String fileName) {
...
+        if (cursor != null && cursor.moveToFirst()) {
+            try {
+                int autoDeleteColumnIndex = cursor.getColumnIndex(FormsProviderAPI.FormsColumns.AUTO_DELETE);
+                autoDelete = cursor.getString(autoDeleteColumnIndex);
+            } finally {
+                cursor.close();
+            }
+        }
+
+        boolean autoDeleteSettings = PreferenceManager.getDefaultSharedPreferences(
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java]: Would be cleaner to use the `GeneralSharedPreferences` abstraction here.

File: collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java [hunk 6]
```diff
@@ -226,6 +228,12 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
 
                 Timber.w("Successfully upgraded database from version %d to %d"
                         + ", without destroying all the old data", oldVersion, newVersion);
+            } else {
+                db.execSQL("ALTER TABLE " + FORMS_TABLE_NAME + " ADD COLUMN "
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/provider/FormsProvider.java]: You'll want to see whether the column already exists before adding it. See #1000 for a description of why and a solution used for the instance table.

**Review by lognaturel — CHANGES_REQUESTED** [review]

**Comment by grzesiek2010** [comment]
@lognaturel Done. Please review again.

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
Code looks good! I'll let @mmarciniak90 do some verification.

**Comment by grzesiek2010** [comment]
@lognaturel 

> Code looks good! I'll let @mmarciniak90 do some verification.

I've already created a ticket for her in Assembla as I thought you may be too busy.

**Comment by mmarciniak90** [comment]
@grzesiek2010  **I noticed problem when I send finalized forms to google spreadsheet.** 

Here is a example of use case to reproduce:
1. Auto Send is Off, Delete After send is Off in ODK
2. <submission auto-delete="true" auto-submit="true" /> in form
3. Server Type: Google Drive, Google Sheets
4. User finishes form 
5. Form is auto sent but form is not removed. User can open form in 'View Sent Form'. Crossed eye icon is NOT visible.

I tested few configurations when ODK Aggregate was set as a server. All cases work as expected.

Tested settings configurations:
1. Test case:
- Auto Send - Off
- Delete after send - Off
- auto-delete="true"
- auto-submit="true"

Result: Form is auto sent and removed from device. There is no opportunity to open sent form in 'View Sent Form'. Crossed eye icon is visible.

2. Test case:
- Auto Send - On
- Delete after send - On
- auto-delete="false"
- auto-submit="false"

Result: Form need to be manually selected to send. User can open form in 'View Sent Form'. Crossed eye icon is NOT visible.

3. Test case:
- Auto Send - On
- Delete after send - Off
- auto-delete="true"
- auto-submit="false"

Result: Form need to be manually selected to send. There is no opportunity to open sent form in 'View Sent Form'. Crossed eye icon is visible.

4. Test case:
- Auto Send - Off
- Delete after send - On
- auto-delete="false"
- auto-submit="true"

Result: Form is auto sent. User can open form in 'View Sent Form'. Crossed eye icon is NOT visible.

Tested with 'Don't keep activities' turned on and off.
Tested on Android 4.1, 4.4, 6.0

**Comment by grzesiek2010** [comment]
@mmarciniak90 I've fixed that issue but I've decided to refactor the code so please test everything again carefully.

@lognaturel I could just fix the issue changing 10-15 lines but the previous approach was really wrong so I've decided to refactor it and now we use deleting functionality in one place.

**Comment by mmarciniak90** [comment]
Tested with success 4 test cases described in previous comment. 
All tests passed with expected result.
Tested on Android 6.0, 5.1, 4.2, 4.1
Verified ODK Aggregate and Google Sheets platforms.
Tested with 'Don't keep activities' turned on and off.

File: collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java [hunk 7]
```diff
@@ -528,25 +524,15 @@ public void authRequest(Uri url, HashMap<String, String> doneSoFar) {
...
-            JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
-            sheetsService = new com.google.api.services.sheets.v4.Sheets.Builder(
-                    transport, jsonFactory, credential)
-                    .setApplicationName("ODK-Collect")
-                    .build();
-            driveService = new com.google.api.services.drive.Drive.Builder(
-                    transport, jsonFactory, credential)
-                    .setApplicationName("ODK-Collect")
-                    .build();
+    private class InstanceGoogleSheetsInstanceUploaderTask extends InstanceGoogleSheetsUploader {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java]: Can this be called `InstanceGoogleSheetsUploaderTask` instead? It seems odd to me to have `Instance` in the name twice!

**Review by lognaturel — COMMENTED** [review]
It is very nice to see that redundancy between the uploaders captured. I think you did the right thing with refactoring here. I made a small naming comment inline. Thanks for separating the changes into meaningful commits, that made the review easier to do.

@mmarciniak90 We were talking earlier about putting together a master manual testing document. The tests you've described here are a great example that could be added as they are, I think. Add a link to this pull request and a note about verifying both Google Sheets upload and Aggregate upload and I think it will be a useful resource.

File: collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java (hunk 7 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/GoogleSheetsUploaderActivity.java]: Fixed.
